types/views/views.go

```diff
@@ -181,3 +185,90 @@ func (v IPPrefixSlice) MarshalJSON() ([]byte, error) {
 func (v *IPPrefixSlice) UnmarshalJSON(b []byte) error {
 	return v.ж.UnmarshalJSON(b)
 }
+
+// MapOf returns a read-only view over m.
+func MapOf[K comparable, V comparable](m map[K]V) Map[K, V] {
+	return Map[K, V]{m}
+}
+
+// Map is a read-only accessor over a map whose values are immutable.
+type Map[K comparable, V any] struct {
+	// ж is the underlying mutable value, named with a hard-to-type
+	// character that looks pointy like a pointer.
+	// It is named distinctively to make you think of how dangerous it is to escape
+	// to callers. You must not let callers be able to mutate it.
+	ж map[K]V
+}
+
+// IsNil reports whether the underlying map is nil.
+func (m Map[K, V]) IsNil() bool {
+	return m.ж == nil
+}
+
+// Len returns the number of elements in the map.
+func (m Map[K, V]) Len() int { return len(m.ж) }
+
+// Get returns the element with key k.
+func (m Map[K, V]) Get(k K) V {
+	return m.ж[k]
+}
+
+// GetOk returns the element with key k and a bool representing whether the key
+// is in map.
+func (m Map[K, V]) GetOk(k K) (V, bool) {
+	v, ok := m.ж[k]
+	return v, ok
+}
+
+// ForEach calls f for every k,v pair in the underlying map.
+func (m Map[K, V]) ForEach(f func(k K, v V)) {
+	for k, v := range m.ж {
+		f(k, v)
+	}
+}
+
+// MapFnOf returns a MapFn for m.
+func MapFnOf[K comparable, T any, V any](m map[K]T, f func(T) V) MapFn[K, T, V] {
+	return MapFn[K, T, V]{
+		ж:     m,
+		wrapv: f,
+	}
+}
+
+// MapFn is like Map but with a func to convert values from T to V.
+// It is used to provide map of slices and views.
+type MapFn[K comparable, T any, V any] struct {
```

Comment:
You already have a `ViewCloner` constraint, I think we should make that less specific on structs and more general purpose.

When I was exploring this problem, I came up with: https://go.dev/play/p/Kn5HUXEWMwe

Reply:
In my snippet, my `Viewer` constraint analogous to your `ViewCloner` constraint.